File: android-app/feature/feature-discover/src/main/java/tv/bayit/plus/feature/discover/walkthrough/WalkthroughSessionManager.kt
package tv.bayit.plus.feature.discover.walkthrough

import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow
import tv.bayit.plus.core.common.logging.BayitLogger
import javax.inject.Inject
import javax.inject.Singleton

@Singleton
class WalkthroughSessionManager @Inject constructor(
    private val logger: BayitLogger,
) {
    private val _activeSession = MutableStateFlow<WalkthroughSession?>(null)
    val activeSession: StateFlow<WalkthroughSession?> = _activeSession.asStateFlow()

    private var stateMachine: WalkthroughStateMachine? = null

    val isActive: Boolean get() = _activeSession.value != null

    val sessionToken: String? get() = _activeSession.value?.sessionToken

    val currentFeatureId: String? get() = _activeSession.value?.featureId

    fun currentStateMachine(): WalkthroughStateMachine? = stateMachine

    fun start(session: WalkthroughSession) {
        if (_activeSession.value != null) {
            logger.warning(
                "walkthrough_replacing_session",
                mapOf(
                    "previous" to (_activeSession.value?.featureId ?: ""),
                    "new" to session.featureId,
                ),
            )
            end()
        }
        stateMachine = WalkthroughStateMachine(session.feature)
        _activeSession.value = session
        logger.info(
            "walkthrough_session_started",
            mapOf("featureId" to session.featureId, "token" to session.sessionToken),
        )
    }

    fun end() {
        val session = _activeSession.value ?: return
        stateMachine?.skip()
        logger.info(
            "walkthrough_session_ended",
            mapOf("featureId" to session.featureId),
        )
        stateMachine = null
        _activeSession.value = null
    }
}

File: ios-app/Packages/BayitCore/Sources/BayitCore/Discover/WalkthroughSessionManager.swift
import Foundation
import Observation
#if canImport(UIKit)
    import UIKit
#endif

@Observable
@MainActor
public final class WalkthroughSessionManager {
    public static let shared = WalkthroughSessionManager()

    public private(set) var activeSession: WalkthroughSession?
    public var onSessionEnd: ((String, Int, Bool) async -> Void)?

    public var isActive: Bool {
        activeSession != nil
    }

    public var sessionToken: String? {
        activeSession?.sessionToken
    }

    public var currentFeatureId: String? {
        activeSession?.featureId
    }

    public var stateMachine: WalkthroughStateMachine? {
        activeSession?.stateMachine
    }

    private let logger = BayitLogger(category: "WalkthroughSession")

    private init() {
        #if canImport(UIKit) && !os(watchOS)
            observeBackgroundTransition()
        #endif
    }

    public func start(session: WalkthroughSession) {
        if let existing = activeSession {
            logger.warning(
                "Ending existing walkthrough before starting new one",
                context: ["previousFeature": existing.featureId, "newFeature": session.featureId]
            )
            end()
        }
        activeSession = session
        logger.info(
            "Walkthrough session started",
            context: ["featureId": session.featureId]
        )
    }

    public func end() {
        guard let session = activeSession else { return }
        let featureId = session.featureId
        let steps = session.stateMachine.totalSteps
        let skipped = !session.stateMachine.isComplete
        if session.stateMachine.isActive {
            session.stateMachine.skip()
        }
        logger.info(
            "Walkthrough session ended",
            context: ["featureId": featureId]
        )
        activeSession = nil
        if let callback = onSessionEnd {
            Task { await callback(featureId, steps, skipped) }
        }
    }

    #if canImport(UIKit) && !os(watchOS)
        private func observeBackgroundTransition() {
            NotificationCenter.default.addObserver(
                forName: UIApplication.didEnterBackgroundNotification,
                object: nil,
                queue: .main
            ) { [weak self] _ in
                Task { @MainActor in
                    self?.end()
                }
            }
        }
    #endif
}
